torch/csrc/utils/python_ptr.h

```diff
@@ -0,0 +1,127 @@
+#include <torch/csrc/utils/python_compat.h>
+#include <stdexcept>
+
+namespace torch::pyptr {
+
+struct Owned;
+
+/// Represents a borrowed PyObject*. The object won't incref'd on acquire or
+/// decref'd on release.
+struct Borrowed {
+  ~Borrowed() = default;
+  Borrowed() : m_value(nullptr) {}
+  Borrowed(PyObject* p) : m_value(p) {}
+  Borrowed(const Borrowed& o) = default;
+  Borrowed& operator=(const Borrowed& o) = default;
+
+  Borrowed(Borrowed&& o) = delete;
+  Borrowed& operator=(Borrowed&& o) = delete;
+
+  static Borrowed none() {
+    return Borrowed(Py_None);
+  }
+
+  static Borrowed false_() {
+    return Borrowed(Py_False);
+  }
+
+  static Borrowed true_() {
+    return Borrowed(Py_True);
+  }
+
+  PyObject* ptr() const {
+    return m_value;
+  }
+
+  Py_hash_t hash() const {
+    return PyObject_Hash(m_value);
+  }
+
+  bool equal(Borrowed other) const {
+    auto result = PyObject_RichCompareBool(m_value, other.m_value, Py_EQ);
+    switch (result) {
+      case 0:
+        return false;
+      case 1:
+        return true;
+      default:
+        throw std::runtime_error("PyObject_RichCompareBool failed");
+    }
+  }
+
+  /// Take ownership of the borrowed reference.
+  Owned to_owned() const;
+
+ protected:
+  PyObject* m_value;
+};
+
+/// Represents an owned PyObject*. You must be explicit on acquire (either with
+/// Owned::own() or Owned::steal()). Releases the
+/// reference on release.
+struct Owned : Borrowed {
```

Comment:
We have some similar classes already in PyTorch.  For example `THPObjectPtr`.  There are also the pybind classes that do the same thing and are used in PyTorch.  PyBind is slow for doing actually binding and C++/Python conversions, but I think the object/handle stuff should be fine.

Can we reuse the existing things?

Reply:
I originally wrote this using pybind. Doing it this way was measurably faster. I looked around for something that already existed but didn't find THPObjectPtr (hard to figure out what to search for). I'll switch to using that.